src/libbson/src/bson/bson-dsl.h

```diff
@@ -0,0 +1,1336 @@
+ * pointer.
+ */
+#define bsonBuild(BSON, ...)                                              \
+   _bsonDSL_begin ("Build a new document for '%s'", _bsonDSL_str (BSON)); \
+   bson_t *_bbDest = &(BSON);                                             \
+   bson_init (_bbDest);                                                   \
+   bsonBuildAppend (*_bbDest, __VA_ARGS__);                               \
+   _bsonDSL_end
+
+/**
+ * @brief Declare a variable and build it with the BSON DSL @see bsonBuild
+ */
+#define bsonBuildDecl(Variable, ...)   \
+   bson_t Variable = BSON_INITIALIZER; \
+   bsonBuild (Variable, __VA_ARGS__)
+
+
+struct _bsonBuildContext_t {
+   /// The document that is being built
+   bson_t *doc;
+   /// The key that is pending an append
+   const char *key;
+   /// The length of the string given in 'key'
+   int key_len;
+   /// The index of the array being built (if applicable)
+   int index;
+   /// A buffer for formatting key strings
+   char index_key_str[16];
+   /// The parent context (if building a sub-document)
+   struct _bsonBuildContext_t *parent;
+};
+
+/// A pointer to the current thread's bsonBuild context
+_bson_thread_local_comdat struct _bsonBuildContext_t
+   *_bsonBuildContextThreadLocalPtr = NULL;
+
+struct _bsonVisitContext_t {
+   const bson_t *doc;
+   bson_iter_t iter;
+   const struct _bsonVisitContext_t *parent;
+   int index;
+};
+
+/// A pointer to the current thread's bsonVisit/bsonParse context
+_bson_thread_local_comdat struct _bsonVisitContext_t const
+   *_bsonVisitContextThreadLocalPtr = NULL;
+
+/**
+ * @brief The most recent error from a bsonBuild() DSL command.
+ *
+ * If NULL, no error occurred. Users can assign a value to this string to
+ * indicate failure.
+ */
+_bson_thread_local_comdat const char *bsonBuildError = NULL;
+
+/**
+ * @brief The most recent error from a buildVisit() or bsonParse() DSL command.
+ *
+ * If NULL, no error occurred. Users can assign a value to this string to
+ * indicate an error.
+ *
+ * If this string becomes non-NULL, the current bsonVisit()/bsonParse() will
+ * halt and return.
+ *
+ * Upon entering a new bsonVisit()/bsonParse(), this will be reset to NULL.
+ */
+_bson_thread_local_comdat const char *bsonParseError = NULL;
+
+#define _bsonDSLDebug(...) \
+   _bson_dsl_debug (BSON_DSL_DEBUG, __FILE__, __LINE__, BSON_FUNC, __VA_ARGS__)
+
+
+static BSON_INLINE bool
+_bson_dsl_test_strequal (const char *string, bool case_sensitive)
+{
+   bson_iter_t it = bsonVisitIter;
+   if (bson_iter_type (&it) == BSON_TYPE_UTF8) {
+      uint32_t len;
+      const char *s = bson_iter_utf8 (&it, &len);
+      if (len != (uint32_t) strlen (string)) {
+         return false;
+      }
+      if (case_sensitive) {
+         return memcmp (string, s, len) == 0;
+      } else {
+         return bson_strcasecmp (string, s) == 0;
+      }
+   }
+   return false;
+}
+
+static BSON_INLINE bool
+_bson_dsl_key_is_anyof (const char *key,
+                        const int keylen,
```

Comment:
```suggestion
                        const size_t keylen,
```
`keylen` should have type `size_t` or `uint32_t` (as returned by `bson_iter_key_len`).